src/libraries/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -123,15 +127,20 @@ public static void RegisterFactory(string providerInvariantName, DbProviderFacto
             ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
             ADP.CheckArgumentNull(factory, nameof(factory));
 
-            _registeredFactories[providerInvariantName] = new ProviderRegistration(factory.GetType().AssemblyQualifiedName, factory);
+            string? factoryTypeAssemblyQualifiedName = factory.GetType().AssemblyQualifiedName;
+            if (factoryTypeAssemblyQualifiedName == null)
+            {
+                throw new ArgumentException($"{nameof(factory)} must have an assembly-qualified name");
```

Comment:
I don't think this is necessary.  AssemblyQualifiedName is nullable because Type.AQN will return null if the Type represents a generic type parameter, but that's not the case here.  You should be able to either just assert it's not null or suppress it with a `!`.